File: java/src/test/au/gov/digitalhealth/FindByTermTest.java
package au.gov.digitalhealth;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import java.util.Collection;
import java.util.HashSet;

import org.junit.Test;

import au.gov.digitalhealth.db.ConceptFinder;
import au.gov.digitalhealth.model.Concept;

/**
 * Test searching for concepts by textual term (description) 
 */
public class FindByTermTest {

    /** Concept: Punch drunk */
    private static final long KNOWN_INACTIVE_CONCEPT_ID = 51996004L;
    
    /** Concept: Punch drunk */
    private static final long KNOWN_ACTIVE_CONCEPT_ID = 230283005L;

    /**
     * Search for the term "drunk". This has two concepts with the same description (a duplication) however
     * one is inactive - so the test is that the search result should contain the active concept but not the
     * inactive concept.
     */
    @Test
    public void testKnownTerm() {
        Collection<Concept> results = ConceptFinder.findByTerm("drunk");
        
        assertNotNull("Expected an empty collection, not null reference", results);
        assertFalse("Known term not found", results.size() == 0);
        assertTrue("Expected more than one concept found for known term", results.size() > 1);
        
        boolean foundActive = false;
        boolean foundInactive = false;
        
        for (Concept concept : results) {
            foundActive |= (concept.getSctId() == KNOWN_ACTIVE_CONCEPT_ID);
            foundInactive |= (concept.getSctId() == KNOWN_INACTIVE_CONCEPT_ID);            
        }
        
        assertTrue("Expected active concept was not present in results", foundActive);
        assertFalse("Inactive concept was present in the result, should only contain active concepts", foundInactive);
    }
    
    @Test
    public void testUnknownTerm() {
        Collection<Concept> results = ConceptFinder.findByTerm("wakawaka");
        
        assertNotNull("Expected an empty collection, not null reference", results);
        assertTrue("Expected no concepts for known invalid term", results.size() == 0);
    }
    
    @Test
    public void testNoDuplicates() {
        HashSet<Long> uniqueIds = new HashSet<Long>();
        Collection<Concept> results = ConceptFinder.findByTerm("heart");
        
        for (Concept c : results) {
            Long sctId = c.getSctId();
            assertFalse("Duplicate concept with SCTID " + sctId + " returned in result", uniqueIds.contains(sctId));
            uniqueIds.add(sctId);
        }
    }
    
    @Test
    public void testMaxRows() {
        int resultLimit = ConceptFinder.getResultLimit();
        Collection<Concept> results = ConceptFinder.findByTerm("heart");
        assertFalse("Number of results returned exceeds the maximum row count specified", results.size() > resultLimit);
    }
}

File: dotNet/UnitTest/FindByTermTest.cs
/** Copyright (c) 2018 Australian Digital Health Agency
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using System.Collections.Generic;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace CTDemo.UnitTest
{
    /// <summary>
    /// Test searching for concepts by textual term (description) 
    /// </summary>
    [TestClass]
    public class FindByTermTest
    {

    // Concept: Punch drunk 
    private const long KNOWN_INACTIVE_CONCEPT_ID = 51996004L;
    
    // Concept: Punch drunk 
    private const long KNOWN_ACTIVE_CONCEPT_ID = 230283005L;
 
     /// <summary>
     /// Search for the term "drunk". This has two concepts with the same description (a duplication) however
     /// one is inactive - so the test is that the search result should contain the active concept but not the
     /// inactive concept.
     /// <summary>
     [TestMethod]
    public void TestKnownTerm() {
        List<Concept> results = ConceptFinder.FindByTerm("drunk");
        
        Assert.IsNotNull(results,"Expected an empty collection, not null reference");
        Assert.IsFalse(results.Count == 0, "Known term not found");
        Assert.IsTrue(results.Count > 1,"Expected more than one concept found for known term");
        
        bool foundActive = false;
        bool foundInactive = false;

        foreach (Concept concept in results)
        {
            foundActive |= (concept.sctId == KNOWN_ACTIVE_CONCEPT_ID);
            foundInactive |= (concept.sctId == KNOWN_INACTIVE_CONCEPT_ID);            
        }

        Assert.IsTrue(foundActive, "Expected active concept was not present in results");
        Assert.IsFalse(foundInactive, "Inactive concept was present in the result, should only contain active concepts");
    }
    

     [TestMethod]
    public void TestUnknownTerm() {
        List<Concept> results = ConceptFinder.FindByTerm("wakawaka");
        
        Assert.IsNotNull(results,"Expected an empty collection, not null reference");
        Assert.IsTrue(results.Count == 0,"Expected no concepts for known invalid term");
    }
    
     [TestMethod]
    public void TestNoDuplicates() {
        HashSet<long> uniqueIds = new HashSet<long>();

        List<Concept> results = ConceptFinder.FindByTerm("heart");
        
        foreach (Concept c in results) {
            long sctId = c.sctId;
            Assert.IsFalse(uniqueIds.Contains(sctId),"Duplicate concept with SCTID " + sctId.ToString() + " returned in result");
            uniqueIds.Add(sctId);
        }
    }

     [TestMethod]
    public void TestMaxRows() {
        int resultLimit = DataSource.GetMaxRows();
        List<Concept> results = ConceptFinder.FindByTerm("heart");
        Assert.IsFalse(results.Count > resultLimit,"Number of results returned exceeds the maximum row count specified");
    }


    }
}
